examples/read_write_mutex/source/app.d

```diff
@@ -0,0 +1,88 @@
+
+import vibe.d;
+
+interface MyBlockingRestInterface {
+	@path("/")
+	int getIndex();
+}
+
+class RestInterfaceImplementation : MyBlockingRestInterface {
+	private {
+		TaskReadWriteMutex m_mutex;
+	}
+
+	this()
+	{
+		m_mutex = new TaskReadWriteMutex(TaskReadWriteMutex.Policy.PREFER_WRITERS);
+	}
+
+	//This method naively simulates access to a shared resource protected
+	//by a TaskReadWriteMutex. Read operations are performed recently (95%) 
+	//and finish quickly (1ms wait time), while write operations occur rarely
+	//(5%) and take much longer to process (50ms wait time).
+	int getIndex()
+	{
+	    import std.random;
+	    import std.stdio;
+	    auto i = uniform(0,20);
+	    if (i == 1) //In a rare case, lock for writing
+	    {
+	        synchronized(m_mutex.writer)
+	        {
+	        	version(PrintRequests)
+	        		writeln("Blocked for writing!");
+	        	//Simulate a slow operation
+	            vibe.core.core.sleep(50.msecs);
+	        }
+	    }
+	    else //More regularly, lock for reading
+	    {
+	        synchronized(m_mutex.reader)
+	        {
+	        	version(PrintRequests)
+	        		writeln("Blocked for reading!");
+	        	//Simulate a faster operation
+	            vibe.core.core.sleep(1.msecs);
+	        }
+	    }
+	    return 0;
+	}
+}
+
+shared static this()
+{
+	auto routes = new URLRouter;
+	registerRestInterface(routes, new RestInterfaceImplementation());
+
+	auto settings = new HTTPServerSettings();
+	settings.port = 8080;
+	settings.options = HTTPServerOption.parseURL | HTTPServerOption.distribute;
+	settings.bindAddresses = ["::1", "127.0.0.1"];
+
+	listenHTTP(settings, routes);
+
+	//Wait for a second for the server to be initialized properly and then start
+	//multiple concurrent threads that simultaneously start queries on the rest
+	//interface defined above.
+	setTimer(1.seconds, {
+		import std.parallelism : totalCPUs;
+
+		for (int cpu = 0; cpu < totalCPUs; ++cpu)
+		{
+			auto val = async({
+					auto api = new RestInterfaceClient!MyBlockingRestInterface("http://127.0.0.1:8080");
+
+					for (int i = 0; i < 100000; ++i)
+						api.getIndex();
+
+					//Since we can't join threads in a setTimer call,
+					//exit the event loop after the first thread has complete all
+					//of it's requests. In any other code, this is propably a bad idea.
+					scope(exit)
+						exitEventLoop(true);
```

Comment:
This is still a really unfortunate limitation currently, I really need to bite the bullet and implement that. A clean workaround would be to use a `TaskCondition` that gets notified at the end of the `async` callback. Probably `Future` should in fact also just use that internally until `Task.join()` is supported across threads.

`RestInterfaceClient`, if the `requestFilter` property isn't modified after initialization, should be thread-safe in practice, because all fields are *de-facto* immutable. I wouldn't do that in an example, though, before this use case is actually supported by the type system. It would be an interesting addition to mirror all methods as `shared` and to add a `shared` constructor. On the other hand, constructing a new `RestInterfaceClient` is cheap and doing it in a module constructor once per thread would eliminate the overhead anyway.